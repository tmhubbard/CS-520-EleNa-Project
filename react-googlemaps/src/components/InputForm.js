import React from 'react';
import Slider from './Slider';
import PlacesAutocomplete, { geocodeByAddress, getLatLng } from 'react-places-autocomplete';



class InputForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleStartAutocompleteChange = this.handleStartAutocompleteChange.bind(this);
        this.handleStartSelect = this.handleStartSelect.bind(this);
        this.handleEndAutocompleteChange = this.handleEndAutocompleteChange.bind(this);
        this.handleEndSelect = this.handleEndSelect.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleSliderChange = this.handleSliderChange.bind(this);
    }

    handleStartAutocompleteChange = startAddress => {
        this.props.onStartChange(null, startAddress);
    };

    handleStartSelect = async value => {
        const results = await geocodeByAddress(value);
        const latLng = await getLatLng(results[0]);
        this.props.onStartChange(latLng, value);
    };

    handleEndAutocompleteChange = endAddress => {
        this.props.onEndChange(null, endAddress);
    };

    handleEndSelect = async value => {
        const results = await geocodeByAddress(value);
        const latLng = await getLatLng(results[0]);
        this.props.onEndChange(latLng, value);
    };
    
    handleChange(event) {
        const target = event.target;
        const value = target.value;
        const name = target.name;
        if(name === "elevationType"){
            this.props.onTypeChange(value);
        }
    }

    handleSubmit(event) {
        event.preventDefault(); //prevents form from actually being submitted
        this.props.submit();
    }
    
    handleSliderChange(percent) {
        this.setState({percentRoute: percent});
        this.props.onPercentChange(percent);
    }

    render() {      
        return (
            <div //style={{float: 'left', width: '25%', height: '100%'}}
            >
            <form onSubmit = {this.handleSubmit}>
                <label style={{fontWeight: "bold"}}>
                    Start Point
                    <br />
                    <PlacesAutocomplete
                        value={this.props.startAddress}
                        onChange={this.handleStartAutocompleteChange}
                        onSelect={this.handleStartSelect}
                    >
                        {({ getInputProps, suggestions, getSuggestionItemProps, loading }) => (
                            <div>
                                <input
                                {...getInputProps({
                                    placeholder: 'Search Places ...',
                                    className: 'location-search-input',
                                })}
                                />
                                <div className="autocomplete-dropdown-container">
                                {loading && <div>Loading...</div>}
                                {suggestions.map(suggestion => {
                                    const className = suggestion.active
                                    ? 'suggestion-item--active'
                                    : 'suggestion-item';
                                    // inline style for demonstration purpose
                                    const style = suggestion.active
                                    ? { backgroundColor: '#fafafa', cursor: 'pointer' }
                                    : { backgroundColor: '#ffffff', cursor: 'pointer' };

                                    return (
                                    <div
                                        {...getSuggestionItemProps(suggestion, {
                                        className,
                                        style,
                                        })}
                                    >
                                        <span>{suggestion.description}</span>
                                    </div>
                                    );
                                })}
                                </div>
                            </div>
                            )}
                    </PlacesAutocomplete>
                </label>
                <br />
                <label style={{fontWeight: "bold"}}>
                    End Point
                    <br />
                    <PlacesAutocomplete
                        value={this.props.endAddress}
                        onChange={this.handleEndAutocompleteChange}
                        onSelect={this.handleEndSelect}
                    >
                        {({ getInputProps, suggestions, getSuggestionItemProps, loading }) => (
                            <div>
                                <input
                                {...getInputProps({
                                    placeholder: 'Search Places ...',
                                    className: 'location-search-input',
                                })}
                                />
                                <div className="autocomplete-dropdown-container">
                                {loading && <div>Loading...</div>}
                                {suggestions.map(suggestion => {
                                    const className = suggestion.active
                                    ? 'suggestion-item--active'
                                    : 'suggestion-item';
                                    // inline style for demonstration purpose
                                    const style = suggestion.active
                                    ? { backgroundColor: '#fafafa', cursor: 'pointer' }
                                    : { backgroundColor: '#ffffff', cursor: 'pointer' };

                                    return (
                                    <div
                                        {...getSuggestionItemProps(suggestion, {
                                        className,
                                        style,
                                        })}
                                    >
                                        <span>{suggestion.description}</span>
                                    </div>
                                    );
                                })}
                                </div>
                            </div>
                            )}
                    </PlacesAutocomplete>
                </label>
                <br />
                <label style={{fontWeight: "bold"}}>
                    Type of Route:
                </label>
                <div style={{display: "block"}}>
                    <input type="radio" id="min" name="elevationType" value="min" checked="checked" onChange={this.handleChange}/>
                    <label htmlFor="min">Minimum Elevation</label>
                    <br />
                    <input type="radio" id="max" name="elevationType" value="max" onChange={this.handleChange}/>
                    <label htmlFor="max">Maximum Elevation</label>
                </div>
                <br />
                <label style={{fontWeight: "bold"}}>
                    Maximum Distance:
                </label>
                <div className="slidecontainer">
                    <Slider onSliderChange = {this.handleSliderChange}/>
                </div>
                <input type="submit" value="Find Route" style={{marginLeft: "20px"}}/>
            </form>
            </div>
        );
    }
}

export default InputForm;