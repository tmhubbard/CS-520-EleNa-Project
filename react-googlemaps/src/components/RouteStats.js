import React from 'react';

class RouteStats extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div style={{float: 'left', clear: 'both'}}>
                <label>
                    Total Elevation Gain:
                    <p>
                        {this.props.elevation} meters
                    </p>
                </label>
                <label>
                    Total Distance:
                    <p>
                        {this.props.distance} meters
                    </p>
                </label>
            </div>
        );
    }
}

export default RouteStats;