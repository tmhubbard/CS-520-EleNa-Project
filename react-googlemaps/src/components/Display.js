import React from 'react';
import InputForm from './InputForm';
import MapComponent from './MapComponent';
import RouteStats from './RouteStats';

class Display extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            mapcenter: {lat: 42.3732, lng: -72.5199},
            route: null,
            renderRoute: false,
            isStartingMarkerShown: false,
            isEndMarkerShown: false,
            startPoint: null,
            endPoint: null,
            elevationType: "min",
            percentRoute: 10,
            startAddress: '',
            endAddress: '',
            totalElevation: 0,
            totalDistance:  0
        };

        this.handleMapStartChange = this.handleMapStartChange.bind(this);
        this.handleMapEndChange = this.handleMapEndChange.bind(this);
        this.handleInputStartChange = this.handleInputStartChange.bind(this);
        this.handleInputEndChange = this.handleInputEndChange.bind(this);
        this.handleInputTypeChange = this.handleInputTypeChange.bind(this);
        this.handleInputPercentChange = this.handleInputPercentChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }
    //callback functions
    handleMapStartChange(location, address) {
        this.setState({isStartingMarkerShown: true, startPoint : location, startAddress: address, mapCenter: location});
        // console.log(location);
        // console.log(address);
    }

    handleMapEndChange(location, address) {
        this.setState({isEndMarkerShown: true, endPoint: location, endAddress: address, mapcenter: location});
        // console.log(location);
        // console.log(address);
    }

    handleInputStartChange(location, address) {
        if(location != null) {
            this.setState({isStartingMarkerShown: true, startPoint: location, startAddress: address});
        }
        else {
            this.setState({startPoint: location, startAddress: address});
        }
        // console.log(location);
        // console.log(address);
    }

    handleInputEndChange(location, address) {
        if(location != null) {
            this.setState({isEndMarkerShown: true, endPoint : location, endAddress: address});
        }
        else {
            this.setState({endPoint : location, endAddress: address});
        }
        // console.log(location);
        // console.log(address);
    }

    handleInputTypeChange(type) {
        this.setState({elevationType: type});
        // console.log(type);
    }

    handleInputPercentChange(percent) {
        this.setState({percentRoute: percent});
        // console.log(percent);
    }

    // handleRecenter(location) {
    //     this.setState({mapcenter: location});
    //     console.log(location);
    // }

    handleSubmit() {
        var submission = {
            start_point: this.state.startPoint,
            end_point: this.state.endPoint,
            elevation_type: this.state.elevationType,
            percent_of_distance: this.state.percentRoute,
        }
        var JSONsubmission = JSON.stringify(submission);
        console.log(JSONsubmission);
        //send data to backend
        //then request data back?

        fetch("http://localhost:5000/getRoute", {
          method: 'POST',
          body: JSONsubmission
        })
        .then(res => res.json())
        .then(json => {
            this.setState({
              route: json.route,
              renderRoute: true,
              totalDistance: json.total_distance_travelled,
              totalElevation: json.total_elevation_gain
            });
            // console.log(json);

            var pathCoordinates = this.state.route;
            console.log(this.state.route);
            var bounds = new window.google.maps.LatLngBounds();
            for (var i = 0; i < pathCoordinates.length; i++) {
                bounds.extend(pathCoordinates[i]);
            }
            var coordinate = {lat: bounds.getCenter().lat(), lng: bounds.getCenter().lng()};
            // console.log(coordinate);
            this.setState({mapcenter: coordinate})
        });
        
        
        
    }


    render() {
        return (
            <div>
                <div style = {{float: 'right', width: '82%', height: '100%', overflowX: 'hidden'}}>
                <MapComponent onStartChange = {this.handleMapStartChange} 
                    onEndChange = {this.handleMapEndChange}
                    startPoint = {this.state.startPoint}
                    endPoint = {this.state.endPoint}
                    // recenter = {this.handleRecenter}
                    isStartingMarkerShown = {this.state.isStartingMarkerShown}
                    isEndMarkerShown = {this.state.isEndMarkerShown}
                    mapcenter = {this.state.mapcenter}
                    renderRoute = {this.state.renderRoute}
                    route = {this.state.route}
                    />
                </div>
                <div style = {{marginLeft: "20px", width: '17%', height: '100%'}}>
                <InputForm onStartChange = {this.handleInputStartChange}
                    onEndChange = {this.handleInputEndChange}
                    onTypeChange = {this.handleInputTypeChange}
                    onPercentChange = {this.handleInputPercentChange}
                    startLocation = {this.state.startPoint}
                    endLocation = {this.state.endPoint}
                    startAddress = {this.state.startAddress}
                    endAddress = {this.state.endAddress}
                    submit = {this.handleSubmit}
                />
                <RouteStats elevation = {this.state.totalElevation}
                    distance = {this.state.totalDistance}/>
                </div>
            </div>
        );
    }
}

export default Display;