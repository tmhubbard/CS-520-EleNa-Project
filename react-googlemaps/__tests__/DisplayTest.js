import React from 'react';
import Display from '../src/components/Display';
import { shallow, mount, render, ReactWrapper } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import InputForm from '../src/components/InputForm';
import RouteStats from '../src/components/RouteStats';

Enzyme.configure({ adapter: new Adapter() });


// ensures Display component renders child components correctly upon initalization
it('renders correctly', () => {
    const result = shallow(<Display />);
    // console.log(result.debug());
    expect(result.debug()).toMatchSnapshot();
});


it('renders GoogleMaps Wrapper', () => {
    const result = shallow(<Display />);
    expect(result.find('Wrapper').exists()).toBe(true);
});


it('renders InputForm and RouteStats', () => {
    const result = shallow(<Display />);
    expect(result.find(InputForm).exists()).toBe(true);
    expect(result.find(RouteStats).exists()).toBe(true);
});


it('renders error header correctly', () => {
    const result = shallow(<Display />);
    expect(result.find('h3').exists()).toBe(false);
    result.find(InputForm).prop('submit')();
    expect(result.find('h3').exists()).toBe(true)
});

it('handleMapStartChange callback function modifies Display state correctly', () => {
    const result = shallow(<Display />);
    expect(result.state('isStartingMarkerShown')).toBe(false);
    expect(result.state('startPoint')).toBe(null);
    expect(result.state('startAddress')).toBe('');
    result.find('Wrapper').prop('onStartChange')({lat: 42.37849519999998, lng: -72.52029559999998}, '39 196 N Pleasant St, Amherst, MA 01002, USA');
    expect(result.state('isStartingMarkerShown')).toBe(true);
    expect(result.state('startPoint')).toStrictEqual({lat: 42.37849519999998, lng: -72.52029559999998});
    expect(result.state('startAddress')).toBe('39 196 N Pleasant St, Amherst, MA 01002, USA');
});

it('handleMapEndChange callback function modifies Display state correctly', () => {
    const result = shallow(<Display />);
    expect(result.state('isEndMarkerShown')).toBe(false);
    expect(result.state('endPoint')).toBe(null);
    expect(result.state('endAddress')).toBe('');
    result.find('Wrapper').prop('onEndChange')({lat: 42.37849519999998, lng: -72.52029559999998}, '39 196 N Pleasant St, Amherst, MA 01002, USA');
    expect(result.state('isEndMarkerShown')).toBe(true);
    expect(result.state('endPoint')).toStrictEqual({lat: 42.37849519999998, lng: -72.52029559999998});
    expect(result.state('endAddress')).toBe('39 196 N Pleasant St, Amherst, MA 01002, USA');
});

it('handleInputStartChange callback function modifies Display state correctly', () => {
    const result = shallow(<Display />);
    expect(result.state('isStartingMarkerShown')).toBe(false);
    expect(result.state('startPoint')).toBe(null);
    expect(result.state('startAddress')).toBe('');
    result.find(InputForm).prop('onStartChange')({lat: 42.37849519999998, lng: -72.52029559999998}, '39 196 N Pleasant St, Amherst, MA 01002, USA');
    expect(result.state('isStartingMarkerShown')).toBe(true);
    expect(result.state('startPoint')).toStrictEqual({lat: 42.37849519999998, lng: -72.52029559999998});
    expect(result.state('startAddress')).toBe('39 196 N Pleasant St, Amherst, MA 01002, USA');
});

it('handleInputEndChange callback function modifies Display state correctly', () => {
    const result = shallow(<Display />);
    expect(result.state('isEndMarkerShown')).toBe(false);
    expect(result.state('endPoint')).toBe(null);
    expect(result.state('endAddress')).toBe('');
    result.find(InputForm).prop('onEndChange')({lat: 42.37849519999998, lng: -72.52029559999998}, '39 196 N Pleasant St, Amherst, MA 01002, USA');
    expect(result.state('isEndMarkerShown')).toBe(true);
    expect(result.state('endPoint')).toStrictEqual({lat: 42.37849519999998, lng: -72.52029559999998});
    expect(result.state('endAddress')).toBe('39 196 N Pleasant St, Amherst, MA 01002, USA');
});


it('handleInputTypeChange callback function modifies Display state correctly', () => {
    const result = shallow(<Display />);
    expect(result.state('elevationType')).toBe("min");
    result.find(InputForm).prop('onTypeChange')("min");
    expect(result.state('elevationType')).toBe("min");
    result.find(InputForm).prop('onTypeChange')("max");
    expect(result.state('elevationType')).toBe("max");
    result.find(InputForm).prop('onTypeChange')("min");
    expect(result.state('elevationType')).toBe("min");
});

it('handleInputPercentChange callback function modifies Display state correctly', () => {
    const result = shallow(<Display />);
    expect(result.state('percentRoute')).toBe(120);
    result.find(InputForm).prop('onPercentChange')(130);
    expect(result.state('percentRoute')).toBe(130);
});


