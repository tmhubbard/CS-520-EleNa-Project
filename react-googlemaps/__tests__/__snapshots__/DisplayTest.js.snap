// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
"<div>
  <div style={{...}}>
    <Wrapper onStartChange={[Function: bound handleMapStartChange]} onEndChange={[Function: bound handleMapEndChange]} startPoint={{...}} endPoint={{...}} isStartingMarkerShown={false} isEndMarkerShown={false} mapcenter={{...}} renderRoute={false} route={{...}} />
  </div>
  <div style={{...}}>
    <InputForm onStartChange={[Function: bound handleInputStartChange]} onEndChange={[Function: bound handleInputEndChange]} onTypeChange={[Function: bound handleInputTypeChange]} onPercentChange={[Function: bound handleInputPercentChange]} startLocation={{...}} endLocation={{...}} startAddress=\\"\\" endAddress=\\"\\" submit={[Function: bound handleSubmit]} />
    <RouteStats elevation={0} distance={0} />
  </div>
</div>"
`;
